### DISCUSSION TITLE
[How-to] Installation in Proxmox LXC container

### DISCUSSION THREAD

**Discussion by jedis00** [mattwebbio/orbital-sync#441] [Show and tell]
I installed orbital-sync in a Proxmox LXC container successfully last night and wanted to share how I accomplished it, while also documenting for myself on how to recreate it when the need arises.

The tutorial will assume you have a basic understanding of Proxmox and Linux.

TODO: Some kind of cron/mechanism to detect if the orbital-sync script dies and restart it. Open to ideas/suggestions here.

- Create a new Debian 12 LXC container in Proxmox.
- The settings I used :
    ```
    arch: amd64
    cores: 1
    features: nesting=1
    hostname: orbitalsync
    memory: 512
    net0: name=eth0,bridge=vmbr0,hwaddr=BC:24:11:xx:xx:xx,ip=dhcp,type=veth
    ostype: debian
    rootfs: ext:vm-100-disk-0,size=4G
    swap: 512
    unprivileged: 1
    ```

Open a shell inside the orbital-sync container to continue the setup instructions.

1. Install updates and package
```
apt update
curl -fsSL https://deb.nodesource.com/setup_20.x | sudo -E bash -
apt-get install -y nodejs curl sudo htop net-tools openssl
npm install -g orbital-sync
```
2. Find location of install
```
$ whereis orbital-sync
/usr/bin/orbital-sync
```
3. Create script to start orbital-sync
nano /root/bin/start.sh
```
#!/bin/bash

VERBOSE=true \
INTERVAL_MINUTES=5 \
PRIMARY_HOST_BASE_URL=https://pihole1.mydomain.net \
PRIMARY_HOST_PATH=/admin \
PRIMARY_HOST_PASSWORD=password \
SECONDARY_HOSTS_1_BASE_URL=https://pihole2.mydomain.net \
SECONDARY_HOSTS_1_PATH=/admin \
SECONDARY_HOSTS_1_PASSWORD=password \
NODE_EXTRA_CA_CERTS="/usr/local/share/ca-certificates/combined_with_root.crt" \
/usr/bin/orbital-sync
```
4. Make script executable
```
chmod +x /root/bin/start.sh
```
5. Get IP address of instance
```
ip a
```
6 Allow root login to scp SSL certs from remote machine
```
nano /etc/ssh/sshd_config
```
7 Add/uncomment this line
```
PermitRootLogin yes
```
8 Restart sshd
```
service sshd restart
```
9 Get certificates ready if pi-hole is operating over SSL. I had to include my Root CA and server cert for it to work.
10. `scp` certificates into the following directory: `/usr/local/share/ca-certificates`
11. Rename cert to have `.crt` suffix. `.pem` was not being detected
```
mv /usr/local/share/ca-certificates/combined_with_root.pem /usr/local/share/ca-certificates/combined_with_root.crt
```
12. Pick up new certs
```
update-ca-certificates
```
13. Set up rc-local to launch orbital-sync at instance start time
```
systemctl status rc-local
nano /etc/rc.local
```
14. Use the following to nohup and launch the script into the background
```
#!/bin/bash

# start orbital-sync
/usr/bin/nohup /root/bin/start.sh &

exit 0
```
15. Make script executable
```
chmod +x /etc/rc.local
```
16. Create rc-local service control file
```
nano /etc/systemd/system/rc-local.service
```
17. Paste the following:
```
[Unit]
 Description=/etc/rc.local Compatibility
 ConditionPathExists=/etc/rc.local

[Service]
 Type=forking
 ExecStart=/etc/rc.local start
 TimeoutSec=0
 StandardOutput=tty
 RemainAfterExit=yes

[Install]
 WantedBy=multi-user.target
```
18. Enable rc.local service and start it
```
systemctl enable rc-local
systemctl start rc-local.service
```
19. Check system status to see that it's running
```
systemctl status rc-local
```
20. Reboot LXC container and validate service is still running:
```
systemctl status rc-local
```

**Comment by coltc50** [comment]
Thank you for the instructions. But I can't get it to work.
I'm hanging on this point 👉 
_9 Get certificates ready if pi-hole is operating over SSL. I had to include my Root CA and server cert for it to work.
10. scp certificates into the following directory: /usr/local/share/ca-certificates
11. Rename cert to have .crt suffix. .pem was not being detected_

Can you explain it a little better what to do? My two piholes are two lxc containers on proxmox. Witch certificate I must copy?
I'm getting this error:
16:27:06 OrbitalSync-Pihole systemd[1]: Started rc-local.service - /etc/rc.local Compatibility.
Oct 05 16:27:08 OrbitalSync-Pihole rc.local[140]: 10/5/2024, 4:27:08 PM: ➡️ Signing in to https://192.168.xx.xx/admin...
Oct 05 16:27:08 OrbitalSync-Pihole rc.local[140]: 10/5/2024, 4:27:08 PM: ⚠ Failure: An unexpected error was thrown:
Oct 05 16:27:08 OrbitalSync-Pihole rc.local[140]: - FetchError: request to https://192.168.xx.xx/admin/index.php?login failed, reason: unable to verify the first certificate
Oct 05 16:27:08 OrbitalSync-Pihole rc.local[140]: 10/5/2024, 4:27:08 PM: Waiting 5 minutes...

**Reply by n8ur** [reply]
Did you ever get an answer to this?  Thanks!

**Comment by Drallas** [comment]
I'm also thinking on doing this, thanks for the guide. 

One question, could you be explicit if there are actions that need to be done on the Pi-hole's during the configuration? 

Or do I have to do everything from the LXC?

**Comment by 64shadow64** [comment]
i can´t get orbital sync to start. Im getting this error:


file:///usr/lib/node_modules/orbital-sync/dist/config/parse.js:68
                throw new MissingRequiredPropertiesError(path, missingRequired);
                      ^

MissingRequiredPropertiesError: Missing required properties for "primaryHost": baseUrl, password
    at parseSchema (file:///usr/lib/node_modules/orbital-sync/dist/config/parse.js:68:23)
    at file:///usr/lib/node_modules/orbital-sync/dist/config/parse.js:61:17
    at Array.map (<anonymous>)
    at parseSchema (file:///usr/lib/node_modules/orbital-sync/dist/config/parse.js:58:62)
    at Config (file:///usr/lib/node_modules/orbital-sync/dist/config/index.js:5:12)
    at file:///usr/lib/node_modules/orbital-sync/dist/index.js:7:16
    at ModuleJob.run (node:internal/modules/esm/module_job:272:25)
    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:552:26)
    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:98:5)

Node.js v23.5.0
